### PR TITLE
[WIP] Better Proguard support

### PR DESCRIPTION
**Pull request by cmelchior**
Annotation processor has been rewritten so all model classes can now be obfuscated fully.
Realm class has been made a lot simpler due to this as well.

Further work:
- Having Realm in both a library project and the main project isn't supported by this. Is there an easy way to do it?
- Consider if compiler and Realm should be split into two projects (afaik only way to get rid of -dontwarn javax.**)
- ~~Consider shading JavaWriter to get rid of Square warning and to reduce future compatibility issues (see https://github.com/google/dagger/issues/79)~~
- Explore possibilities to get RealmProxyMediatorImpl even though it has been obfuscated (possible through annotations).
- ~~Performance test comparing it to previous implementation + if we use HashMap instead of switch.~~

If all 3 things above are implemented we can support proguard without **any** additional configuration.

---------------------

*Edit:* I am reverting to WIP as the current approach will break existing apps that try to have Realm both inside libraries and inside the app.

The setup in this PR needs to be extended to support the following cases:

1) Selfcontained persistence layer inside the lib, ie. app doesn't have to know or care that the lib is using Realm.

2) Lib provides realm model classes to be used by the app.

Some design ideas:
Libraries must be responsible for exposing their modeldata where relevant. Most likely candidate is overloading the Realm-class annotation:
```
@RealmClass(name = "custom-name", exported = false)
```
Think about how to extend the Mediator concept to a recursive data structure provided by libraries:
- How to list number of public types?
- How can the app efficiently get the proper mediator?

Proposal:
Each app/lib does the following:

1) Annotation processor create a Class looking like this:

```
public class RealmPublicModels$<UUID> implements RealmPublicModels {
   @override
   public List<Class<? extends RealmObject>> getPublicModels() {
    return Arrays.asList(TypeA.class, TypeB.class);
  }
}
```

2) A ```realm.properties``` file is created with the classname of the mediator adapter for this jar:
```
mediatorAdapter = RealmMediatorAdapterImpl$<UUID>
```

3) Annotation processor creates an adapter that can find all other adapters and check which adapter should handle a specific RealmModel (expand this)

---------------

Additional notes: Cocoa are also receiving feature requests for controlling Schema's and versions across different Realms. We will must likely need something like that and this PR will need to be aware of it:

https://github.com/realm/realm-cocoa/issues/1336

### CODE REVIEW

**Comment by cmelchior** [comment]
Ready for mege review 

@emanuelez @kneth @bmunkholm 

Note this new approach will remove the ValidationList class which could potentially cause problems for people who are manipulating that file themselves, eg. when working with library projects, see this: https://github.com/realm/realm-java/issues/511

I have some ideas to fix this however, but we can take them offline.

File: examples/introExample/build.gradle [hunk 1]
```diff
@@ -10,13 +10,22 @@ android {
...
         versionName "1.0"
     }
-    buildTypes {
+
+    signingConfigs {
         release {
-            minifyEnabled false
-            //proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
+            storeFile file("keystore/release.keystore")
+            storePassword "realm1234"
```
Comment (kneth) [inline examples/introExample/build.gradle]: Is it safe to store password in a public repository?

Comment (cmelchior) [inline examples/introExample/build.gradle]: It is just a dummy certificate so the password is not secret. Ideally you would probably want to store it in an environment var or something, but doing it like this makes it work out of the box.

File: realm/src/main/java/io/realm/Realm.java [hunk 2]
```diff
@@ -82,6 +83,10 @@
 
     private static final String TAG = "REALM";
     private static final String TABLE_PREFIX = "class_";
+    private static final String INCORRECT_THREAD_MESSAGE = "Realm access from incorrect thread. Realm objects can only be accessed on the thread they were created.";
+    private static final String CLOSED_REALM = "This Realm instance has already been closed, making it unusable.";
```
Comment (kneth) [inline realm/src/main/java/io/realm/Realm.java]: unusable -> inaccessible?

**Comment by emanuelez** [comment]
:+1:

**Comment by bmunkholm** [comment]
remove WIP in subject when it's no longer WIP

File: changelog.txt [hunk 3]
```diff
@@ -4,6 +4,7 @@
  * Adding more logging capabilities at the JNI level.
  * Added proper encryption support. NOTE: The key has been increased from 32 bytes to 64 bytes (see example).
  * Throw NoSuchMethodError when RealmResults.indexOf() is called, since it's not implemented yet.
+ * Improved ProGuard support. Internal ValidationList has been replaced with RealmProxyMediatorImpl.
```
Comment (bmunkholm) [inline changelog.txt]: Better spec which "improvement" has been implemented. What will people experience now?

**Comment by cmelchior** [comment]
Superseded by https://github.com/realm/realm-java/pull/880

**Pull request closed without merging**